agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1504 @@
+                                    } else {
+                                        trace!("do_discover_on_discovery_handler - Discovery Handlers returned error status {}. Marking all instances offline.", status);
+                                        // TODO: Possibly mark config as invalid
+                                        // Mark all associated instances as offline by declaring no visible instances
+                                        discovery_operator
+                                            .update_instance_connectivity_status(
+                                                kube_interface.clone(),
+                                                std::collections::HashMap::new(),
+                                                dh_details.register_request.is_local,
+                                            )
+                                            .await?;
+                                        discovery_operator
+                                            .set_discovery_handler_connectivity_status(
+                                                &endpoint,
+                                                DiscoveryHandlerConnectivityStatus::Online,
+                                            );
+                                        break;
+                                    }
+                                }
+                            }
+                        }
+                        StreamType::Embedded(mut stream) => {
+                            discovery_operator
+                                .internal_do_discover(
+                                    kube_interface.clone(),
+                                    &dh_details,
+                                    &mut stream,
+                                )
+                                .await
+                                .unwrap();
+                            discovery_operator.set_discovery_handler_connectivity_status(
+                                &endpoint,
+                                DiscoveryHandlerConnectivityStatus::Online,
+                            );
+                            break;
+                        }
+                    }
+                }
+                None => {
+                    deregistered = discovery_operator
+                        .mark_offline_or_deregister_discovery_handler(&endpoint)
+                        .await
+                        .unwrap();
+                }
+            }
+            if deregistered {
+                break;
+            } else {
+                // Sleep and keep looping until connection established or deregistered due to grace period elapsing
+                #[cfg(not(test))]
+                tokio::time::delay_for(Duration::from_secs(60)).await;
+                #[cfg(test)]
+                tokio::time::delay_for(Duration::from_millis(100)).await;
+            }
+        }
+        Ok(())
+    }
+}
+
+async fn try_receive<T>(
```

Comment:
is this function needed?

Reply:
I thought it was -- I was having issues with select and the mutability of the receivers but just removed it in  059de92d14 and all is well.